arc_cache.go

```diff
@@ -13,13 +14,20 @@ import (
 type cacheHave bool
 type cacheSize int
 
+type lock struct {
+	mu     sync.RWMutex
+	refcnt int
+}
+
 // arccache wraps a BlockStore with an Adaptive Replacement Cache (ARC) that
 // does not store the actual blocks, just metadata about them: existence and
 // size. This provides block access-time improvements, allowing
 // to short-cut many searches without querying the underlying datastore.
 type arccache struct {
+	lklk sync.Mutex
+	lks  map[string]*lock
```

Comment:
If the allocations here become an issue, we could pool the locks. But really, I doubt that'll make much of a difference in practice.